File: 02.Queue/103.二叉树的锯齿形层序遍历.cpp
/*
 * @lc app=leetcode.cn id=103 lang=cpp
 *
 * [103] 二叉树的锯齿形层序遍历
 *
 * https://leetcode-cn.com/problems/binary-tree-zigzag-level-order-traversal/description/
 *
 * algorithms
 * Medium (57.10%)
 * Likes:    402
 * Dislikes: 0
 * Total Accepted:    117.5K
 * Total Submissions: 205.7K
 * Testcase Example:  '[3,9,20,null,null,15,7]'
 *
 * 给定一个二叉树，返回其节点值的锯齿形层序遍历。（即先从左往右，再从右往左进行下一层遍历，以此类推，层与层之间交替进行）。
 * 
 * 例如：
 * 给定二叉树 [3,9,20,null,null,15,7],
 * 
 * 
 * ⁠   3
 * ⁠  / \
 * ⁠ 9  20
 * ⁠   /  \
 * ⁠  15   7
 * 
 * 
 * 返回锯齿形层序遍历如下：
 * 
 * 
 * [
 * ⁠ [3],
 * ⁠ [20,9],
 * ⁠ [15,7]
 * ]
 * 
 * 
 */

#include <assert.h>
#include <limits.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <algorithm>
#include <iostream>
#include <numeric>
#include <queue>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;



// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
       vector<vector<int>> ans;

        //  生成当前层
        vector<TreeNode*> cur;
        // 注意加入的条件，不能把空指针放进去
        if (root) {
            cur.push_back(root);
        }

        vector<TreeNode*> nextLevel;
        bool bRev = false;

        while (!cur.empty()) {
            // 一定要记得把下一层清空
            nextLevel.clear();

            // 在结果层中新生成一层，ans.back()专门用来存放结果
            ans.emplace_back();

            for (auto c: cur) {
                // 存放当前层的结果
                ans.back().push_back(c->val);

                // 专门用来存放下一层结果
                if (c->left) {
                    nextLevel.push_back(c->left);
                }
                if (c->right) {
                    nextLevel.push_back(c->right);
                }
            }
            if (bRev) {
                reverse(ans.back().begin(), ans.back().end());
            }
            bRev = !bRev;

            // c++里面更迭的技巧
            // 把下一层更迭为当前层。
            cur.swap(nextLevel);
        }

        return ans;
    }
};
// @lc code=end

File: 02.Queue/103.二叉树的锯齿形层序遍历.py
#
# @lc app=leetcode.cn id=103 lang=python
#
# [103] 二叉树的锯齿形层序遍历
#
# https://leetcode-cn.com/problems/binary-tree-zigzag-level-order-traversal/description/
#
# algorithms
# Medium (57.10%)
# Likes:    402
# Dislikes: 0
# Total Accepted:    117.5K
# Total Submissions: 205.7K
# Testcase Example:  '[3,9,20,null,null,15,7]'
#
# 给定一个二叉树，返回其节点值的锯齿形层序遍历。（即先从左往右，再从右往左进行下一层遍历，以此类推，层与层之间交替进行）。
# 
# 例如：
# 给定二叉树 [3,9,20,null,null,15,7],
# 
# 
# ⁠   3
# ⁠  / \
# ⁠ 9  20
# ⁠   /  \
# ⁠  15   7
# 
# 
# 返回锯齿形层序遍历如下：
# 
# 
# [
# ⁠ [3],
# ⁠ [20,9],
# ⁠ [15,7]
# ]
# 
# 
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def zigzagLevelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        # 存放每一层遍历的结果
        ans = []

        # 记录当前层，注意不要把空的root放进去
        cur = []
        if root:
            cur.append(root)
        
        # 当前层需要倒转不？
        bRev = False
        
        while len(cur):
            # nextLevel用来存放下一层的结点
            nextLevel = []

            # curResult用来存放当前层的值
            curResult = []

            for c in cur:
                # 把当前层的值放到curResult里面
                curResult.append(c.val)

                # 生成下一层
                if c.left:
                    nextLevel.append(c.left)
                if c.right:
                    nextLevel.append(c.right)
            
            # 指向下一层
            cur = nextLevel

            # 把当前层放到结果中
            if bRev:
                curResult.reverse()
            ans.append([x for x in curResult])
            bRev = not bRev
        
        # 返回结果
        return ans

# @lc code=end

